### PR TITLE
Document and clean-up CLI parsing code

### PR DESCRIPTION
**Pull request by smondet**
This is a follow up to PR #49.

`make doc` gives:

image

### CODE REVIEW

File: src/lib/EDSL.mli [hunk 1]
```diff
@@ -288,7 +288,58 @@ val tmp_file: ?tmp_dir: string t -> string -> file
...
 
+(** Typed command-line parsing for your shell scripts, à la {!Prtinf.scanf}. *)
 module Command_line: sig
+  (** 
+
+     Use this module like OCaml's {!Printf.scanf} function.
+
+     - Build a command-line “format specification” using the {!Arg} module.
+     - Call the {!parse} function with an appropriately typed function.
+
```
Comment (rleonid) [inline src/lib/EDSL.mli]: I would add, "Here is a potential argument specification for a shell script that downloads and unarchives them."

File: src/lib/EDSL.mli [hunk 2]
```diff
@@ -288,7 +288,58 @@ val tmp_file: ?tmp_dir: string t -> string -> file
...
+     Example: {[
+       let cli_spec =
+         Command_line.Arg.(
+           string
+             ~doc:"The URL to the stuff" ["-u"; "--url"]
+             ~default:no_value
+           & flag ["-c"; "--all-in-tmp"] ~doc:"Do everything in the temp-dir"
```
Comment (rleonid) [inline src/lib/EDSL.mli]: I would get rid of this flag, it is confusing. Where would you operate if not in the temp dir, the current dir? Seems like this variable would be better suited as a string. Perhaps a better argument for a flag (boolean option) would be something like remove intermediate target.

File: src/lib/EDSL.mli [hunk 3]
```diff
@@ -288,7 +288,58 @@ val tmp_file: ?tmp_dir: string t -> string -> file
...
+       (*
+          `cli_spec` has type:
+
+           (string Genspio.EDSL.t ->
+            bool Genspio.EDSL.t ->
+            string Genspio.EDSL.t -> string Genspio.EDSL.t -> unit Genspio.EDSL.t,
+            unit Genspio.EDSL.t)
+           Genspio.EDSL.Command_line.cli_options
+          
+           so the second argument must have type:
```
Comment (rleonid) [inline src/lib/EDSL.mli]: Change to, " so the `action` function (the second argument to `parse`) must have type."

**Comment by smondet** [comment]
Thanks @rleonid 



> "The loop function below is building 3 pieces of code at once:" -> "is building 3 pieces of shell code"

I put “Genspio” code since at that level Shell is a compilation target; not “the code.” :)

> "`applied_action` is a difference list (See: https://drup.github.io/2016/08/02/difflists/) that calls the user provided action function (f) on the parsed arguments. The resulting unit t can then use the parsed command line arguments.

`applied_action` is not the difference list here; it's the `options` argument of
the `parse` function.

`applied_action` is the result of applying `action` to all the elements of
`options` + the list of anonymous arguments.

> I would get rid of this flag, it is confusing. Where would you operate if not
> in the temp dir, the current dir?

I removed it from the doc as you suggested but it's from the actual example: Yes
the default is to extract in the current dir (since it is what you usually
want); but for testing before messing everything up there is the option
`--all-in-tmp`.

**Pull request merged**